tests/acceptance/chartcuterie/test_image_block_builder.py

```diff
@@ -63,13 +65,43 @@ def test_image_block_for_endpoint_regression(self):
             )
         group = Group.objects.first()
         group.update(type=PerformanceP95EndpointRegressionGroupType.type_id)
+        return group
 
+    @with_feature("organizations:slack-endpoint-regression-image")
+    def test_image_block_for_endpoint_regression(self):
+        group = self._create_endpoint_regression_issue()
         with self.feature(self.features):
             image_block = ImageBlockBuilder(group=group).build_image_block()
 
         assert image_block and "type" in image_block and image_block["type"] == "image"
         assert "_media/" in image_block["image_url"]
 
+    @patch("sentry.utils.performance_issues.detectors.utils.escape_transaction")
+    @with_feature("organizations:slack-endpoint-regression-image")
+    def test_caching(self, mock_escape_transaction):
+        mock_escape_transaction.return_value = "Test Transaction"
+        group = self._create_endpoint_regression_issue()
+        image_blocks = []
+
+        def make_request():
+            with self.feature(self.features):
+                image_blocks.append(ImageBlockBuilder(group=group).build_image_block())
+
+        threads = [threading.Thread(target=make_request) for _ in range(2)]
+        # Start all the threads
+        for thread in threads:
+            thread.start()
+
+        # Wait for all threads to finish
+        for thread in threads:
+            thread.join()
+
+        # make_request()
+        # make_request()
```

Comment:
If I comment the thread code and instead call back to back request the result gets returned from the cache but then I won't be properly testing the lock logic.

When using thread though the test completely fail with an exception. It seems like each thread is creating the test DB again 🤔  which I don't understand.

Reply:
Pasting the exception here for convenience:
```
________________________ ERROR at teardown of TestSlackImageBlockBuilder.test_caching ________________________
src/sentry/db/postgres/decorators.py:91: in inner
    return func(self, sql, *args, **kwargs)
src/sentry/db/postgres/base.py:85: in execute
    return self.cursor.execute(sql)
E   psycopg2.errors.ObjectInUse: database "test_region" is being accessed by other users
E   DETAIL:  There are 2 other sessions using the database.

During handling of the above exception, another exception occurred:
.venv/lib/python3.11/site-packages/django/db/backends/utils.py:103: in _execute
    return self.cursor.execute(sql)
src/sentry/db/postgres/decorators.py:77: in inner
    raise_the_exception(self.db, e)
src/sentry/db/postgres/decorators.py:75: in inner
    return func(self, *args, **kwargs)
src/sentry/db/postgres/decorators.py:18: in inner
    return func(self, *args, **kwargs)
src/sentry/db/postgres/decorators.py:93: in inner
    raise type(e)(f"{e!r}\nSQL: {sql}").with_traceback(e.__traceback__)
src/sentry/db/postgres/decorators.py:91: in inner
    return func(self, sql, *args, **kwargs)
src/sentry/db/postgres/base.py:85: in execute
    return self.cursor.execute(sql)
E   psycopg2.errors.ObjectInUse: ObjectInUse('database "test_region" is being accessed by other users\nDETAIL:  There are 2 other sessions using the database.\n')
E   SQL: DROP DATABASE "test_region"

The above exception was the direct cause of the following exception:
.venv/lib/python3.11/site-packages/pytest_django/fixtures.py:150: in django_db_setup
    teardown_databases(db_cfg, verbosity=request.config.option.verbose)
.venv/lib/python3.11/site-packages/django/test/utils.py:366: in teardown_databases
    connection.creation.destroy_test_db(old_name, verbosity, keepdb)
.venv/lib/python3.11/site-packages/django/db/backends/base/creation.py:311: in destroy_test_db
    self._destroy_test_db(test_database_name, verbosity)
.venv/lib/python3.11/site-packages/django/db/backends/base/creation.py:327: in _destroy_test_db
    cursor.execute(
.venv/lib/python3.11/site-packages/sentry_sdk/utils.py:1709: in runner
    return original_function(*args, **kwargs)
.venv/lib/python3.11/site-packages/django/db/backends/utils.py:79: in execute
    return self._execute_with_wrappers(
.venv/lib/python3.11/site-packages/django/db/backends/utils.py:92: in _execute_with_wrappers
    return executor(sql, params, many, context)
.venv/lib/python3.11/site-packages/django/db/backends/utils.py:100: in _execute
    with self.db.wrap_database_errors:
.venv/lib/python3.11/site-packages/django/db/utils.py:91: in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
.venv/lib/python3.11/site-packages/django/db/backends/utils.py:103: in _execute
    return self.cursor.execute(sql)
src/sentry/db/postgres/decorators.py:77: in inner
    raise_the_exception(self.db, e)
src/sentry/db/postgres/decorators.py:75: in inner
    return func(self, *args, **kwargs)
src/sentry/db/postgres/decorators.py:18: in inner
    return func(self, *args, **kwargs)
src/sentry/db/postgres/decorators.py:93: in inner
    raise type(e)(f"{e!r}\nSQL: {sql}").with_traceback(e.__traceback__)
src/sentry/db/postgres/decorators.py:91: in inner
    return func(self, sql, *args, **kwargs)
src/sentry/db/postgres/base.py:85: in execute
    return self.cursor.execute(sql)
E   django.db.utils.OperationalError: ObjectInUse('database "test_region" is being accessed by other users\nDETAIL:  There are 2 other sessions using the database.\n')
E   SQL: DROP DATABASE "test_region"

During handling of the above exception, another exception occurred:
.venv/lib/python3.11/site-packages/pytest_django/fixtures.py:152: in django_db_setup
    request.node.warn(
E   pytest.PytestWarning: Error when trying to teardown test databases: OperationalError('ObjectInUse(\'database "test_region" is being accessed by other users\\nDETAIL:  There are 2 other sessions using the database.\\n\')\nSQL: DROP DATABASE "test_region"')
```

Reply:
Asked @markstory in a separate thread adding response here to close the loop. Django is making new connection in each thread that database is used (which is in every thread here). I'll move to [using existing lock utils](https://github.com/getsentry/sentry/blob/master/src/sentry/utils/locking/lock.py) rather than creating mine from scratch which are better tested and then remove my threads code here.